### PR TITLE
planner: rewrite `LIKE` as range for expression index

### LINKED ISSUE #23912: LIKE not rewritten as range for expression index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create table tbl1 (id int unsigned not null auto_increment primary key, s int, index (( md5(s) )) );
create table tbl2 (id int unsigned not null auto_increment primary key, s char(32), index(s));

insert into tbl1 (id) select null; insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;  insert into tbl1 (id) select null from tbl1;

update tbl1 set s=id%32;

insert into tbl2 (id) select id from tbl1;
update tbl2 set s=md5(id%32);

explain select count(*) from tbl1 where md5(s) like '02e74f10e0327ad868d138f2b4fdd6f0';
explain select count(*) from tbl1 where md5(s) like '02e74f10e0327ad868d138f2b4fdd6f%';
explain select count(*) from tbl1 where md5(s) >= '02e74f10e0327ad868d138f2b4fdd6f' and md5(s) < '02e74f10e0327ad868d138f2b4fdd6g';

explain select count(*) from tbl2 where s like '02e74f10e0327ad868d138f2b4fdd6f0';
explain select count(*) from tbl2 where s like '02e74f10e0327ad868d138f2b4fdd6f%';

### 2. What did you expect to see? (Required)

TiDB correctly rewrites LIKE to be a range scan when reading from a column-based index (the tbl2 example).

TiDB should also be able to rewrite LIKE to a range scan 

### 3. What did you see instead (Required)

```
5.7.25-TiDB-v5.0.0 (root) [test]> explain select count(*) from tbl1 where md5(s) like '02e74f10e0327ad868d138f2b4fdd6f0';
+---------------------------+---------+-----------+----------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| id                        | estRows | task      | access object                                | operator info                                                                                                 |
+---------------------------+---------+-----------+----------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| StreamAgg_10              | 1.00    | root      |                                              | funcs:count(1)->Column#4                                                                                      |
| └─IndexReader_15          | 0.26    | root      |                                              | index:IndexRangeScan_14                                                                                       |
|   └─IndexRangeScan_14     | 0.26    | cop[tikv] | table:tbl1, index:expression_index(md5(`s`)) | range:["02e74f10e0327ad868d138f2b4fdd6f0","02e74f10e0327ad868d138f2b4fdd6f0"], keep order:false, stats:pseudo |
+---------------------------+---------+-----------+----------------------------------------------+---------------------------------------------------------------------------------------------------------------+
3 rows in set (0.001 sec)

5.7.25-TiDB-v5.0.0 (root) [test]> explain select count(*) from tbl1 where md5(s) like '02e74f10e0327ad868d138f2b4fdd6f%';
+-------------------------------+---------+-----------+---------------+----------------------------------------------------------------------+
| id                            | estRows | task      | access object | operator info                                                        |
+-------------------------------+---------+-----------+---------------+----------------------------------------------------------------------+
| StreamAgg_20                  | 1.00    | root      |               | funcs:count(Column#6)->Column#4                                      |
| └─TableReader_21              | 1.00    | root      |               | data:StreamAgg_9                                                     |
|   └─StreamAgg_9               | 1.00    | cop[tikv] |               | funcs:count(1)->Column#6                                             |
|     └─Selection_19            | 204.80  | cop[tikv] |               | like(md5(cast(test.tbl1.s)), "02e74f10e0327ad868d138f2b4fdd6f%", 92) |
|       └─TableRangeScan_18     | 256.00  | cop[tikv] | table:tbl1    | range:[0,+inf], keep order:false, stats:pseudo                       |
+-------------------------------+---------+-----------+---------------+----------------------------------------------------------------------+
5 rows in set (0.001 sec)

5.7.25-TiDB-v5.0.0 (root) [test]> explain select count(*) from tbl1 where md5(s) >= '02e74f10e0327ad868d138f2b4fdd6f' and md5(s) < '02e74f10e0327ad868d138f2b4fdd6g';
+-----------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------+
| id                          | estRows | task      | access object                                | operator info                                                                                               |
+-----------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------+
| StreamAgg_17                | 1.00    | root      |                                              | funcs:count(Column#6)->Column#4                                                                             |
| └─IndexReader_18            | 1.00    | root      |                                              | index:StreamAgg_9                                                                                           |
|   └─StreamAgg_9             | 1.00    | cop[tikv] |                                              | funcs:count(1)->Column#6                                                                                    |
|     └─IndexRangeScan_16     | 6.40    | cop[tikv] | table:tbl1, index:expression_index(md5(`s`)) | range:["02e74f10e0327ad868d138f2b4fdd6f","02e74f10e0327ad868d138f2b4fdd6g"), keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------+
4 rows in set (0.002 sec)

5.7.25-TiDB-v5.0.0 (root) [test]> explain select count(*) from tbl2 where s like '02e74f10e0327ad868d138f2b4fdd6f0';
+---------------------------+---------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------+
| id                        | estRows | task      | access object          | operator info                                                                                                 |
+---------------------------+---------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------+
| StreamAgg_10              | 1.00    | root      |                        | funcs:count(1)->Column#3                                                                                      |
| └─IndexReader_15          | 0.26    | root      |                        | index:IndexRangeScan_14                                                                                       |
|   └─IndexRangeScan_14     | 0.26    | cop[tikv] | table:tbl2, index:s(s) | range:["02e74f10e0327ad868d138f2b4fdd6f0","02e74f10e0327ad868d138f2b4fdd6f0"], keep order:false, stats:pseudo |
+---------------------------+---------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------+
3 rows in set (0.001 sec)

5.7.25-TiDB-v5.0.0 (root) [test]> explain select count(*) from tbl2 where s like '02e74f10e0327ad868d138f2b4fdd6f%';
+-----------------------------+---------+-----------+------------------------+-------------------------------------------------------------------------------------------------------------+
| id                          | estRows | task      | access object          | operator info                                                                                               |
+-----------------------------+---------+-----------+------------------------+-------------------------------------------------------------------------------------------------------------+
| StreamAgg_17                | 1.00    | root      |                        | funcs:count(Column#5)->Column#3                                                                             |
| └─IndexReader_18            | 1.00    | root      |                        | index:StreamAgg_9                                                                                           |
|   └─StreamAgg_9             | 1.00    | cop[tikv] |                        | funcs:count(1)->Column#5                                                                                    |
|     └─IndexRangeScan_16     | 6.40    | cop[tikv] | table:tbl2, index:s(s) | range:["02e74f10e0327ad868d138f2b4fdd6f","02e74f10e0327ad868d138f2b4fdd6g"), keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+------------------------+-------------------------------------------------------------------------------------------------------------+
4 rows in set (0.001 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.0.0
Edition: Community
Git Commit Hash: bdac0885cd11bdf571aad9353bfc24e13554b91c
Git Branch: heads/refs/tags/v5.0.0
UTC Build Time: 2021-04-06 16:39:10
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.002 sec)
```

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #23912 

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note

### CODE REVIEW

File: cmd/explaintest/t/explain_generate_column_substitute.test [hunk 1]
```diff
@@ -174,3 +174,8 @@ explain format = 'brief' select c0 from t0; -- TableRead
 drop table if exists t0;
 create table t0(c0 double, c1 float as (c0) unique);
 explain format = 'brief' select c0 from t0; -- TableRead
+
+create table tbl1 (id int unsigned not null auto_increment primary key, s int, index((md5(s))));
+insert into tbl1 (id) select null; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1; insert into tbl1 (id) select null from tbl1;
+update tbl1 set s=id%32;
+explain select count(*) from tbl1 where md5(s) like '02e74f10e0327ad868d138f2b4fdd6f%';
```
Comment (wjhuang2016) [inline cmd/explaintest/t/explain_generate_column_substitute.test:181]: Can we check the result too?

**Comment by wjhuang2016** [comment]
The title should be:  `planner: ...`

**Review by wjhuang2016 — APPROVED** [review]

**Review by lzmhhh123 — APPROVED** [review]
LGTM

**Pull request merged**